src/Interpreters/InterpreterSetQuery.cpp

```diff
@@ -55,13 +59,40 @@ void InterpreterSetQuery::applySettingsFromQuery(const ASTPtr & ast, ContextMuta
         return;
 
     /// First apply the outermost settings. Then they could be overridden by deeper settings.
-    if (const auto * query_with_output = dynamic_cast<const ASTQueryWithOutput *>(ast.get()))
+    if (const auto * query_with_output = dynamic_cast<ASTQueryWithOutput *>(ast.get()))
     {
         if (query_with_output->settings_ast)
             InterpreterSetQuery(query_with_output->settings_ast, context_).executeForCurrentContext(/* ignore_setting_constraints= */ false);
 
-        if (const auto * create_query = ast->as<ASTCreateQuery>(); create_query && create_query->select)
-            applySettingsFromSelectWithUnion(create_query->select->as<ASTSelectWithUnionQuery &>(), context_);
+        if (const auto * create_query = ast->as<ASTCreateQuery>(); create_query)
+        {
+            if (create_query->select)
+                applySettingsFromSelectWithUnion(create_query->select->as<ASTSelectWithUnionQuery &>(), context_);
+            else if (
+                !create_query->settings_ast && create_query->storage && !create_query->storage->engine && create_query->storage->settings
+                && context_->getApplicationType() != Context::ApplicationType::CLIENT)
+            {
+                /// If we parsed one set of settings and no engine we don't know if they belong to the query
+                /// or to the engine. To disambiguate, we try to apply them in the context, if it works then we move
+                /// them there; if it doesn't we assume they are engine settings.
+                /// Note that this only works in the server, since the client ignores unknown settings (for compatibility reasons)
```

Comment:
There's still some ambiguity:
```
 :) select name from system.settings where name in (select name from system.merge_tree_settings)

SELECT name
FROM system.settings
WHERE name IN (
    SELECT name
    FROM system.merge_tree_settings
)

Query id: 02123176-93ed-44ca-96a6-2dca88094d00

    ┌─name────────────────────────────────────────────────────────┐
 1. │ min_compress_block_size                                     │
 2. │ max_compress_block_size                                     │
 3. │ allow_suspicious_indices                                    │
 4. │ parts_to_delay_insert                                       │
 5. │ parts_to_throw_insert                                       │
 6. │ number_of_mutations_to_delay                                │
 7. │ number_of_mutations_to_throw                                │
 8. │ max_partitions_to_read                                      │
 9. │ async_insert                                                │
10. │ merge_tree_clear_old_temporary_directories_interval_seconds │
11. │ merge_tree_clear_old_parts_interval_seconds                 │
12. │ lightweight_mutation_projection_mode                        │
    └─────────────────────────────────────────────────────────────┘

12 rows in set. Elapsed: 0.017 sec. Processed 1.36 thousand rows, 295.91 KB (82.18 thousand rows/s., 17.91 MB/s.)
Peak memory usage: 7.34 KiB.

```

Reply:
I've flipped the checks to remove the ambiguity in a clearer way and support other use cases. Now, if there is only one SETTING declaration (2 of them is unambiguous already) we check all the settings declared to see if they are supported by the storage, and if not check if they are global settings to move them there.

This supports mixing the 2 types of settings together: e.g `ENGINE MergeTree() ORDER BY tuple() SETTINGS allow_experimental_json_type=1, replace_long_file_name_to_hash=1` will move the JSON setting to the context (and allow creating a table with JSON in that statement) while keeping the MergeTree setting in the definition.

I'm currently running CI to see what breaks and will add more tests to verify things work correctly